lib/fog/core/cache.rb

```diff
@@ -0,0 +1,89 @@
+module Fog
+  class Cache
+
+    class CacheNotFound < StandardError; end
+
+    require 'fileutils'
+    require 'yaml'
+
+    attr_reader :model
+
+    def self.load(model_klass, service)
+      cache_files = Dir.glob("#{sandbox(model_klass, service)}/*")
+
+      if cache_files.empty?
+        raise CacheNotFound
+      end
+
+      new_instances = cache_files.map { |path|
+        model_klass = cache_data(path)[:model_klass]
+        new_instance = model_klass.new(cache_data(path)[:attrs])
+      }
+
+      collection_klass = cache_data(cache_files.sample)[:collection_klass]
+
+      # set service & collection so model has api connectivity/logic.
+      # https://github.com/fog/fog-aws/issues/354#issuecomment-286789702
+      new_instances.each { |i|
+        i.instance_variable_set(:@service, service) if service
+        i.collection = collection_klass.new({service: service}) if collection_klass
+      }
+
+    end
+
+    def self.expire_cache!(model_klass, service)
+      FileUtils.rm_rf(sandbox(model_klass, service))
+    end
+
+    def self.cache_data(path)
+      @cache_data ||= {}
+      return @cache_data[path] if @cache_data[path]
+      @cache_data[path] = YAML.load(File.read(path))
+    end
+
+    def self.sandbox(model_klass, service)
+      if service.nil?
+        name = "/tmp/#{safe_path(model_klass)}"
+      else
+
+        # XXXX
+        # TODO ... how to get an effective service identifier?
+        # XXXX
+
+        service_identifier = service.region
```

Comment:
The service identifier (which needs refining as to how to identify it across various Fog constructs) needs to be embedded into the cache identifier:

If you use different region or AWS account and the cache is not expired, when you load via cache it can assume incorrectly that any model klass caches (for example, an AWS Server class) belongs to that service even though it belongs to a different one.